src/logicTemplates/limitorder.js

```diff
@@ -0,0 +1,108 @@
+const templates = require('./templates');
+const algosdk = require('../main');
+const logicSig = require('../logicsig');
+
+class LimitOrder {
+    /**
+     * MakeLimitOrder allows a user to exchange some number of assets for some number of algos.
+     * Fund the contract with some number of Algos to limit the maximum number of
+     * Algos you're willing to trade for some other asset.
+     *
+     * Works on two cases:
+     * * trading Algos for some other asset
+     * * closing out Algos back to the originator after a timeout
+     *
+     * trade case, a 2 transaction group:
+     * gtxn[0] (this txn) Algos from Me to Other
+     * gtxn[1] asset from Other to Me
+     *
+     * We want to get _at least_ some amount of the other asset per our Algos
+     * gtxn[1].AssetAmount / gtxn[0].Amount >= N / D
+     * ===
+     * gtxn[1].AssetAmount * D >= gtxn[0].Amount * N
+     *
+     * close-out case:
+     * txn alone, close out value after timeout
+     * Constructor Parameters:
+     * @param {string} owner: the address to refund funds to on timeout
+     * @param {int} assetid: the ID of the transferred asset
+     * @param {int} ratn: exchange rate (N asset per D Algos, or better)
+     * @param {int} ratd: exchange rate (N asset per D Algos, or better)
+     * @param {int} expiryRound: the round at which the account expires
+     * @param {int} minTrade: the minimum amount (of Algos) to be traded away
+     * @param {int} maxFee: maximum fee used by the limit order transaction
+     * @returns {LimitOrder}
+     */
+    constructor(owner, assetid, ratn, ratd, expiryRound, minTrade, maxFee) {
+        // don't need to validate owner - it will be validated by template.insert
+        if (!Number.isSafeInteger(assetid) || assetid < 0) throw Error("assetid must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(ratn) || ratn < 0) throw Error("ratn must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(ratd) || ratd < 0) throw Error("ratd must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(expiryRound) || expiryRound < 0) throw Error("expiryRound must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(minTrade) || minTrade < 0) throw Error("minTrade must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(maxFee) || maxFee < 0) throw Error("maxFee must be a positive number and smaller than 2^53-1");
+
```

Comment:
Could these validation tests, where I see in other scripts as well, be pulled out into a helper?

Reply:
:+1: Good thinking, will file a ticket